File: mobile/ios/ShiftPlanner/ShiftPlanner/Features/Manager/Employees/Repositories/EmployeeManagementRepository.swift
import Foundation

struct EmployeeManagementCapabilities {
    let canCreatePosition: Bool
    let canAssignPosition: Bool
    let canRemovePosition: Bool
    let canRemoveEmployee: Bool
}

protocol EmployeeManagementRepository {
    var capabilities: EmployeeManagementCapabilities { get }

    func fetchEmployees() async throws -> [ManagedEmployee]
    func fetchBranches() async throws -> [ManagedBranch]
    func fetchPositions() async throws -> [ManagedPosition]
    func addPosition(title: String, currentPositions: [ManagedPosition]) async throws -> [ManagedPosition]
    func removePosition(_ position: ManagedPosition, from employees: [ManagedEmployee], positions: [ManagedPosition]) async throws -> EmployeeManagementSnapshot
    func assignBranch(_ branchId: Int?, to employee: ManagedEmployee, in employees: [ManagedEmployee]) async throws -> [ManagedEmployee]
    func assignPosition(_ positionId: Int?, to employee: ManagedEmployee, in employees: [ManagedEmployee]) async throws -> [ManagedEmployee]
    func removeEmployee(_ employee: ManagedEmployee, from employees: [ManagedEmployee]) async throws -> [ManagedEmployee]

    // MARK: - Depth: linking, work limits, join requests

    /// Links an existing user (by their 16-char public ID) into the company.
    func linkEmployeeByPublicId(publicId: String, branchId: Int?, positionId: Int?) async throws -> ManagedEmployee

    func fetchWorkLimits(employeeId: Int) async throws -> WorkLimits
    func updateWorkLimits(employeeId: Int, maxHoursPerWeek: Int, maxHoursPerDay: Int) async throws -> WorkLimits

    func fetchManagerRequests() async throws -> [PendingManagerRequest]
    func acceptManagerRequest(id: Int) async throws
    func declineManagerRequest(id: Int) async throws

    func fetchEmployeeRequests() async throws -> [PendingEmployeeRequest]
    func acceptEmployeeRequest(id: Int) async throws
    func declineEmployeeRequest(id: Int) async throws
}

struct EmployeeManagementSnapshot {
    let employees: [ManagedEmployee]
    let positions: [ManagedPosition]
}

File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/data/employees/EmployeeManagementRepository.kt
package com.froggyriia.shiftplanner.data.employees

import com.froggyriia.shiftplanner.domain.model.ManagedBranch
import com.froggyriia.shiftplanner.domain.model.ManagedBranchAssignment
import com.froggyriia.shiftplanner.domain.model.ManagedEmployee
import com.froggyriia.shiftplanner.domain.model.ManagedEmployeeCalendar
import com.froggyriia.shiftplanner.domain.model.ManagedPosition
import com.froggyriia.shiftplanner.domain.model.PendingEmployeeRequest
import com.froggyriia.shiftplanner.domain.model.PendingManagerRequest
import com.froggyriia.shiftplanner.domain.model.WorkLimits

interface EmployeeManagementRepository {
    suspend fun fetchEmployees(): List<ManagedEmployee>
    suspend fun fetchBranches(): List<ManagedBranch>
    suspend fun fetchPositions(): List<ManagedPosition>
    suspend fun createEmployee(
        fullName: String,
        email: String,
        positionId: Int,
        branchId: Int?
    ): ManagedEmployee
    suspend fun deleteEmployee(employeeId: Int)
    suspend fun assignPosition(employeeId: Int, positionId: Int?): ManagedEmployee
    suspend fun assignBranch(employeeId: Int, branchId: Int?): ManagedEmployee
    suspend fun createPosition(title: String, companyId: Int): ManagedPosition
    suspend fun deletePosition(positionId: Int)

    suspend fun linkEmployeeByPublicId(publicId: String, branchId: Int?, positionId: Int?): ManagedEmployee

    // ── Multi-branch, calendar, add-manager ───────────────────────────────────
    suspend fun addManagerByPublicId(publicId: String): PendingManagerRequest
    suspend fun fetchEmployeeBranches(employeeId: Int): List<ManagedBranchAssignment>
    suspend fun replaceEmployeeBranches(employeeId: Int, branchIds: List<Int>, primaryBranchId: Int): List<ManagedBranchAssignment>
    suspend fun fetchEmployeeCalendar(employeeId: Int): ManagedEmployeeCalendar

    // ── Work-hours limits ─────────────────────────────────────────────────────
    suspend fun fetchWorkLimits(employeeId: Int): WorkLimits
    suspend fun updateWorkLimits(employeeId: Int, maxHoursPerWeek: Int, maxHoursPerDay: Int): WorkLimits

    // ── Pending join requests ─────────────────────────────────────────────────
    suspend fun fetchManagerRequests(): List<PendingManagerRequest>
    suspend fun acceptManagerRequest(id: Int)
    suspend fun declineManagerRequest(id: Int)
    suspend fun fetchEmployeeRequests(): List<PendingEmployeeRequest>
    suspend fun acceptEmployeeRequest(id: Int)
    suspend fun declineEmployeeRequest(id: Int)
}
